doc/decisions/header_file_structure.md

```diff
@@ -12,31 +12,135 @@ The big problem is kdbprivate.h. It has two main problems:
-    Here a local header could be used as well (requires longer relative includes), or we could create a src/include/testing folder (could be added to include path).
-    Symbols belonging to this category should only appear in the private section of the symbols.map file (we could also create a second private section, if is just there for testing, or adjust some compiler configs for testing).
-3.  Then there are things that I would call internal or pseudo-private. This includes the struct \_Key and struct \_KeySet, but also keyClearSync or elektraMalloc and the high-level functions needed for codegen. These are things that shouldn't be part of the public API, because we don't want to make all the guarantees associated with that, but nonetheless cannot truly be private. In other words, each of these functions/structs/symbols has a specific "target audience" that may use it.
-    Here I would either create a separate kdbinternal.h or just keep kdbprivate.h, as long as it is a separate file from 4.
-    Symbols belonging to this category should only appear in the private section of the symbols.map file.
-4.  Finally, there is what I'd call unstable or pseudo-public. This includes e.g. ksFindHierarchy or elektraReadArrayNumber. These functions could be public, but for various reasons are not. Maybe they are not well-tested or maybe we just don't want to commit to the function yet. Many of these (but not all) can't be part of libelektra-ease, because the rely on something more private (e.g. directly access the array in a KeySet).
-    Here I would either create a separate kdbinternal.h or just keep kdbprivate.h, as long as it is a separate file from 3.
-    Symbols belonging to this category should only appear in the private section of the symbols.map file.
+### Libraries
+
+A library can be monolithic or modularized.
+Monolithic libraries should be small and bigger libraries with large APIs should be modularized.
+
+A monolithic library `foo` may have these headers (covering categories 3 & 4 from above):
+
+- `src/include/elektra/foo.h`:
+  Contains the full public API of `libelektra-foo`.
+  Will be installed as `<include-root>/elektra/foo.h`.
+- `src/include/internal/foo.h`:
+  Contains the internal API of `libelektra-foo`.
+  Will not be installed.
+
+A modularized library `foo` may have these headers (covering categories 3 & 4 from above):
+
+- `src/include/elektra/foo.h`:
+  Declares the public API of `libelektra-foo`, by including `#include <elektra/foo/*.h>`.
+  Will be installed as `<include-root>/elektra/foo/public.h`.
+
+  Such a header may only contain `#include <elektra/foo/*.h>` lines.
+
+- `src/include/elektra/foo/*.h`:
+  Additional public API header of `libelektra-foo`.
+  Will be installed as `<include-root>/elektra/foo/*.h`.
+
+  All of these installed headers `installed.h` must be included from `src/include/elektra/foo.h` via a line `#include <elektra/foo/installed.h>`.
```

Comment:
Sentence is broken

Reply:
I don't see the problem...

Would this make it better:

```suggestion
  All of these "installed headers" (`installed.h`) must be included from `src/include/elektra/foo.h` via a line `#include <elektra/foo/installed.h>`.
```

Reply:
It is a bit better but still: "installed headers" is plural, `installed.h` is one header. Maybe simply a "for example" is missing to fix the sentence but I am not sure what you want to say here at the moment.

Reply:
How about 

```suggestion
  Any one `installed.h` of these installed headers must be included from `src/include/elektra/foo.h` via a line `#include <elektra/foo/installed.h>`.
```

Reply:
Yes, better.